### PR TITLE
Beef up BufRead::consume docs

### LINKED ISSUE #23196: BufReader .consume() confuses the Reader seek point when the buffer is empty
The behavior of .consume when the buffer is empty is weird : 

```rust
let mut f = BufReader::new(Cursor::new(vec![65, 255, 42]));
f.consume(1);
let x = &mut [0; 1];
println!("{}", f.read(x).unwrap()); // prints 0
```
while : 

```rust
let mut f = BufReader::new(Cursor::new(vec![65, 255, 42]));
let x = &mut [0; 1];
f.read(x).unwrap();
f.consume(1);
println!("{}", f.read(x).unwrap()); // prints 1, x[0] is 42
```

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #23196

### CODE REVIEW

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -558,6 +558,10 @@ pub trait BufRead: Read {
     /// This function does not perform any I/O, it simply informs this object
     /// that some amount of its buffer, returned from `fill_buf`, has been
     /// consumed and should no longer be returned.
+    ///
+    /// This function is used to tell the buffer how many bytes you've consumed from the return
+    /// value of `fill_buff`, and so may do odd things if `fill_buf` isn't called before calling
```
Comment (sfackler) [inline src/libstd/io/mod.rs]: `buff`

**Comment by oli-obk** [comment]
also there's no mention that the number of bytes passed to `consume` must be `<=` to the number of bytes in the buffer returned by `fill_buf`

**Comment by steveklabnik** [comment]
@oli-obk thanks! fixed. That naming convention is something I have yet to internalize, ugh.

File: src/libstd/io/mod.rs [hunk 2]
```diff
@@ -558,6 +558,12 @@ pub trait BufRead: Read {
     /// This function does not perform any I/O, it simply informs this object
     /// that some amount of its buffer, returned from `fill_buf`, has been
     /// consumed and should no longer be returned.
+    ///
+    /// This function is used to tell the buffer how many bytes you've consumed from the return
+    /// value of `fill_buf`, and so may do odd things if `fill_buf` isn't called before calling
+    /// this.
```
Comment (alexcrichton) [inline src/libstd/io/mod.rs]: Can this follow the surrounding style of wrapping to 80 characters? Other than that r=me

**Pull request merged**